enhancements/ingress/request-timeout-configuration.md

```diff
@@ -0,0 +1,231 @@
+---
+title: request-timeout-configuration
+authors:
+  - "@rfredette"
+reviewers:
+  - "@miciah"
+  - "@danehans"
+  - "@frobware"
+  - "@sgreene570"
+  - "@knobunc"
+  - "@miheer"
+  - "@candita"
+  - "@alebedev87"
+approvers:
+  - "@miciah"
+  - "@frobware"
+  - "@knobunc"
+creation-date: 2021-06-17
+last-updated: 2021-06-28
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced|informational
+see-also:
+replaces:
+superseded-by:
+---
+
+# Request Timeout Configuration
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This document proposes the addition of several API fields into the ingress
+controller API, allowing the cluster admin to adjust how long routers will hold
+open connections that are awaiting a response.
+
+## Motivation
+
+Customers need the ability to adjust how long HAProxy holds connections open,
+either by extending the timeout to accomodate slower backends or clients, or by
+shortening the timeout, allowing connections to be closed more aggressively.
+
+### Goals
+
+Allow admins to configure the following HAProxy connection timeout parameters:
+- timeout client
+- timeout client-fin
+- timeout server
+- timeout server-fin
+- tcp-request inspect-delay
+
+### Non-Goals
+
+- Allow admins to configure every HAProxy timeout parameter
+
+## Proposal
+
+Add the following configurable variables to the ingresscontroller API:
+```yaml
+spec:
+  tuningOptions:
+    clientTimeout: "30s"
+    clientFinTimeout: "1s"
+    serverTimeout: "30s"
+    serverFinTimeout: "1s"
+    tlsInspectDelay: "10s"
+```
+
+All variables expect time values. If no units are specified, the router assumes
+milliseconds.
```

Comment:
The Go definition should use the `metav1.Duration` type, which is parsed using `time.ParseDuration`:
```go
// Duration is a wrapper around time.Duration which supports correct
// marshaling to YAML and JSON. In particular, it marshals into strings, which
// can be used as map keys in json.
type Duration struct {
	time.Duration `protobuf:"varint,1,opt,name=duration,casttype=time.Duration"`
}
```
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/duration.go#L24-L29

If you use `metav1.Duration`, the unit suffix is required.